File: diff_redactor.py
"""Redact sensitive patterns from Git diffs before sending to the frontend."""

import re

_PLACEHOLDER = "[REDACTED]"

# Patterns that match sensitive values; replacement preserves structure for diff readability
_REDACT_PATTERNS: list[tuple[re.Pattern, str]] = [
    # OpenRouter / OpenAI / generic API keys (sk-or-..., sk-...)
    (re.compile(r'\bsk-or-[a-zA-Z0-9_-]{20,}\b'), "sk-or-[REDACTED]"),
    (re.compile(r'\bsk-[a-zA-Z0-9_-]{20,}\b'), "sk-[REDACTED]"),
    # AWS access keys
    (re.compile(r'\bAKIA[0-9A-Z]{16}\b'), _PLACEHOLDER),
    # Generic "key=value" or "KEY=value" (env-style)
    (re.compile(r'(\b(?:api_key|apikey|secret|password|passwd|token|credential)\s*=\s*["\']?)([^"\'\s]+)(["\']?)', re.I), r"\1[REDACTED]\3"),
    # Connection strings with credentials
    (re.compile(r'(://)([^:@]+):([^@]+)(@)'), r"\1\2:[REDACTED]\4"),
    # Bearer tokens
    (re.compile(r'\bBearer\s+[a-zA-Z0-9_\-\.]+', re.I), "Bearer [REDACTED]"),
    # Generic hex-like tokens (40+ chars, e.g. SHA1/SHA256 hashes used as secrets)
    (re.compile(r'\b[a-fA-F0-9]{40,}\b'), _PLACEHOLDER),
]


def redact_diff(diff: str) -> str:
    """
    Redact sensitive patterns from a Git diff string.
    Returns the redacted diff. Empty input returns empty string.
    """
    if not diff or not isinstance(diff, str):
        return diff if diff else ""
    result = diff
    for pattern, replacement in _REDACT_PATTERNS:
        try:
            result = pattern.sub(replacement, result)
        except re.error:
            continue
    return result

File: webui/src/lib/diff-redactor.ts
const PLACEHOLDER = "[REDACTED]";

const REDACT_PATTERNS: Array<[RegExp, string]> = [
  [/\bsk-or-[a-zA-Z0-9_-]{20,}\b/g, "sk-or-[REDACTED]"],
  [/\bsk-[a-zA-Z0-9_-]{20,}\b/g, "sk-[REDACTED]"],
  [/\bAKIA[0-9A-Z]{16}\b/g, PLACEHOLDER],
  [/(\b(?:api_key|apikey|secret|password|passwd|token|credential)\s*=\s*["']?)([^"'\s]+)(["']?)/gi, "$1[REDACTED]$3"],
  [/(:\/\/)([^:@]+):([^@]+)(@)/g, "$1$2:[REDACTED]$4"],
  [/\bBearer\s+[a-zA-Z0-9_\-\.]+/gi, "Bearer [REDACTED]"],
  [/\b[a-fA-F0-9]{40,}\b/g, PLACEHOLDER],
];

export function redactDiff(diff: string): string {
  if (!diff || typeof diff !== "string") return diff || "";
  let result = diff;
  for (const [pattern, replacement] of REDACT_PATTERNS) {
    try {
      result = result.replace(pattern, replacement);
    } catch {
      // skip invalid patterns
    }
  }
  return result;
}
